### ISSUE TITLE
Multi-cluster: Proxy calls external Ingress GW directly instead of via Egress GW

### ISSUE THREAD

**Issue by bzurkowski** [istio/istio#22244]
**Bug description**

**tl;dr**
1. Envoy proxy in  *sleep* pod attempts to call external Ingress Gateway directly instead of routing request via the Egress Gateway in the first place. It calls the Egress Gateway as a fallback after 10 second connect timeout.
2. Egress Gateway (or other remote service on the request path) responds with HTTP 400 Bad Request on TLS Client Hello.

Deployed multi-cluster Istio service mesh spanned across two clusters (Kubernetes 1.16, Istio 1.3.5) by following the [Replicated Control Planes](https://istio.io/docs/setup/install/multicluster/gateways/) guide in the documentation.

Application pods deployed in the mesh do not have direct access to the external network, therefore egress traffic must be routed through the Istio Egress Gateway. I used the following ServiceEntry policy to enforce that behaviour:

```yaml
Name:         httpbin-bar
Namespace:    foo
Labels:       <none>
API Version:  networking.istio.io/v1alpha3
Kind:         ServiceEntry
Metadata:
  Creation Timestamp:  2020-03-13T12:11:34Z
  Generation:          1
  Resource Version:    47682708
  Self Link:           /apis/networking.istio.io/v1alpha3/namespaces/foo/serviceentries/httpbin-bar
  UID:                 10bb9f03-01ec-4f50-abd8-d8538222f04a
Spec:
  Addresses:
    240.0.0.2
  Endpoints:
    Address:  101.102.114.161
    Network:  external
    Ports:
      http1:  15443
    Address:  10.100.1.253
    Ports:
      http1:  15443
  Hosts:
    httpbin.bar.global
  Location:  MESH_INTERNAL
  Ports:
    Name:      http1
    Number:    8000
    Protocol:  http
  Resolution:  STATIC
Events:        <none>
```
When attempting to call the *httpbin* service from *sleep* pod, I get `503 Service Unavailable` response. It happens always with 10 second delay, which is Envoy's default for the connect timeout.

I ran *tcpdump* session inside network namespace of *sleep* pod:
```
10.100.2.153.55600 -> 101.102.114.161.15433	TCP	[SYN]
10.100.2.153.55600 -> 101.102.114.161.15433	TCP	[SYN] - retransmission
10.100.2.153.55600 -> 101.102.114.161.15433	TCP	[SYN] - retransmission

10.100.2.153.34764 -> 10.100.2.253.15433	TCP	[SYN]
10.100.2.253.15433 -> 10.100.2.153.34764	TCP	[SYN, ACK]
10.100.2.153.34764 -> 10.100.2.253.15433	TCP	[ACK]
10.100.2.153.34764 -> 10.100.2.253.15433	TLSv1	[Client Hello]
10.100.2.253.15433 -> 10.100.2.153.34764	TCP	[ACK]
10.100.2.253.15433 -> 10.100.2.153.34764	HTTP1.1	[400 Bad Request]
10.100.2.153.34764 -> 10.100.2.253.15433	TCP	[ACK]
10.100.2.153.34764 -> 10.100.2.253.15433	TCP	[PSH, ACK]
10.100.2.153.34764 -> 10.100.2.253.15433	TCP	[RST, ACK]
```

It turns out, before routing the request through the Egress Gateway, Envoy attempts (for 10 seconds) to connect to the destination IP address in the external network (`101.102.114.161`). Only after connect timeout, it routes the request via the Egress Gateway, which fails to complete the request anyway due to failed TLS handshake with the remote proxy.

To sum up, there are two related issues at the moment:

1. Envoy proxy in  *sleep* pod attempts to call external Ingress Gateway directly instead of routing request via the Egress Gateway in the first place. It calls the Egress Gateway as a fallback after 10 second connect timeout.
2. Egress Gateway (or other service on the request path) responds with HTTP 400 Bad Request on TLS Client Hello.

Pods and services in Cluster 1 (*foo* namespace):
```
NAME                     TYPE           CLUSTER-IP     EXTERNAL-IP       PORT(S)                                                                                                                                      AGE
istio-citadel            ClusterIP      10.100.1.162   <none>            8060/TCP,15014/TCP                                                                                                                           4d4h
istio-egressgateway      ClusterIP      10.100.1.253   <none>            80/TCP,443/TCP,15443/TCP                                                                                                                     4d4h
istio-galley             ClusterIP      10.100.1.242   <none>            443/TCP,15014/TCP,9901/TCP                                                                                                                   4d4h
istio-ingressgateway     LoadBalancer   10.100.1.202   101.102.114.211   15020:30475/TCP,80:31380/TCP,443:31390/TCP,31400:31400/TCP,15029:30708/TCP,15030:30326/TCP,15031:31486/TCP,15032:31788/TCP,15443:30249/TCP   4d4h
istio-pilot              ClusterIP      10.100.1.38    <none>            15010/TCP,15011/TCP,8080/TCP,15014/TCP                                                                                                       4d4h
istio-policy             ClusterIP      10.100.1.70    <none>            9091/TCP,15004/TCP,15014/TCP                                                                                                                 4d4h
istio-sidecar-injector   ClusterIP      10.100.1.241   <none>            443/TCP,15014/TCP                                                                                                                            4d4h
istio-telemetry          ClusterIP      10.100.1.158   <none>            9091/TCP,15004/TCP,15014/TCP,42422/TCP                                                                                                       4d4h
istiocoredns             ClusterIP      10.100.1.120   <none>            53/UDP,53/TCP                                                                                                                                4d4h
prometheus               ClusterIP      10.100.1.2     <none>            9090/TCP 
```
```
NAME                                                          READY   STATUS      RESTARTS   AGE    IP             NODE
istio-citadel-994d884d9-q8hdn                                 1/1     Running     0          4d4h   10.100.4.88    node2
istio-egressgateway-79b6467b8-zfmrj                           1/1     Running     0          2d4h   10.100.3.147   node2
istio-galley-6797695d7f-wnzrp                                 1/1     Running     0          2d4h   10.100.3.134   node2
istio-ingressgateway-55898ccb8-rvrnh                          1/1     Running     0          2d4h   10.100.2.199   node2
istio-init-crd-10-1.3.5-kxpq2                                 0/1     Completed   0          4d4h   10.100.2.104   node2
istio-init-crd-12-1.3.5-qbzsk                                 0/1     Completed   0          4d4h   10.100.2.117   node2
istio-pilot-cf4dfcf58-wskmx                                   2/2     Running     0          4d     10.100.3.163   node3
istio-policy-5bc47fcd5-dwr4t                                  2/2     Running     0          2d4h   10.100.5.238   node2
istio-sidecar-injector-5d65cfcd79-q9xgv                       1/1     Running     0          4d4h   10.100.4.217   node2
istio-telemetry-b88b4d794-gww2p                               2/2     Running     3          4d4h   10.100.4.248   node3
istiocoredns-576dd6459c-84ffr                                 2/2     Running     0          4d4h   10.100.4.111   node2
prometheus-6f74d6f76d-wm4zd                                   1/1     Running     0          4d4h   10.100.3.19    node2
```
```
NAME                                                          READY   STATUS    RESTARTS   AGE    IP             NODE
sleep-f8cbf5b76-tj82b                                         2/2     Running   0          2d4h   10.100.2.153   node2
```

Pods and services in Cluster 2 (*bar* namespace):
```
NAME                     TYPE           CLUSTER-IP     EXTERNAL-IP       PORT(S)                                                                                                                                      AGE
istio-citadel            ClusterIP      10.100.1.247   <none>            8060/TCP,15014/TCP                                                                                                                           4d4h
istio-egressgateway      ClusterIP      10.100.1.92    <none>            80/TCP,443/TCP,15443/TCP                                                                                                                     4d4h
istio-galley             ClusterIP      10.100.1.197   <none>            443/TCP,15014/TCP,9901/TCP                                                                                                                   4d4h
istio-ingressgateway     LoadBalancer   10.100.1.236   101.102.114.161   15020:30942/TCP,80:31380/TCP,443:31390/TCP,31400:31400/TCP,15029:31259/TCP,15030:32648/TCP,15031:30240/TCP,15032:31604/TCP,15443:31262/TCP   4d4h
istio-pilot              ClusterIP      10.100.1.180   <none>            15010/TCP,15011/TCP,8080/TCP,15014/TCP                                                                                                       4d4h
istio-policy             ClusterIP      10.100.1.15    <none>            9091/TCP,15004/TCP,15014/TCP                                                                                                                 4d4h
istio-sidecar-injector   ClusterIP      10.100.1.182   <none>            443/TCP,15014/TCP                                                                                                                            4d4h
istio-telemetry          ClusterIP      10.100.1.217   <none>            9091/TCP,15004/TCP,15014/TCP,42422/TCP                                                                                                       4d4h
istiocoredns             ClusterIP      10.100.1.35    <none>            53/UDP,53/TCP                                                                                                                                4d4h
prometheus               ClusterIP      10.100.1.240   <none>            9090/TCP                                                                                                                                     4d4h
```
```
NAME                                                          READY   STATUS      RESTARTS   AGE     IP             NODE
istio-citadel-994d884d9-tftks                                 1/1     Running     0          4d4h    10.100.2.148   node2
istio-egressgateway-79799969b4-4jzd4                          1/1     Running     0          4d4h    10.100.5.154   node1
istio-galley-6797695d7f-8vgz6                                 1/1     Running     0          4d4h    10.100.5.167   node1
istio-ingressgateway-55898ccb8-c6xr2                          1/1     Running     0          4d4h    10.100.5.62    node2
istio-init-crd-10-1.3.5-q2fn7                                 0/1     Completed   0          4d4h    10.100.5.200   node1
istio-init-crd-11-1.3.5-p6l6t                                 0/1     Completed   0          4d4h    10.100.5.52    node1
istio-init-crd-12-1.3.5-cdgk6                                 0/1     Completed   0          4d4h    10.100.3.57    node2
istio-pilot-cf4dfcf58-qd8xc                                   2/2     Running     0          4d4h    10.100.5.179   node1
istio-policy-5bc47fcd5-v2zjp                                  2/2     Running     3          4d4h    10.100.4.205   node1
istio-sidecar-injector-5d65cfcd79-68lt2                       1/1     Running     0          4d4h    10.100.5.120   node2
istio-telemetry-b88b4d794-2jcx7                               2/2     Running     2          4d4h    10.100.2.67    node2
istiocoredns-576dd6459c-kfvgr                                 2/2     Running     0          4d4h    10.100.4.95    node2
prometheus-6f74d6f76d-t89mt                                   1/1     Running     0          4d4h    10.100.2.207   node1
```
```
NAME                                                          READY   STATUS    RESTARTS   AGE   IP             NODE
httpbin-779c54bf49-88pwn                                      2/2     Running   0          4d    10.100.5.184   node1
```

**Expected behavior**

HTTP request initiated from *sleep* pod completes with 200 response status code.

**Steps to reproduce the bug**

Just follow the [Replicated Control Planes](https://istio.io/docs/setup/install/multicluster/gateways/) guide in the documentation.

**Version**

Kubernetes:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-18T14:36:53Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-18T14:27:17Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```

Istio:
```
client version: 1.3.5
control plane version: 1.3.5
```

Helm:
```
Client: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
```

**How was Istio installed?**

Using Istio Helm charts.

**Environment where bug was observed (cloud vendor, OS, etc)**

RedHat 7.4, Kubernetes 1.16, Istio 1.3.5

**Comment by bzurkowski** [comment]
Inspected proxy endpoints for *sleep* pod and Egress Gateway:
```
$ istioctl -n foo proxy-config endpoints sleep-f8cbf5b76-lhh9c |grep httpbin
10.100.1.253:15443        HEALTHY     OK	outbound|8000||httpbin.bar.global
101.102.114.161:15443     HEALTHY     OK	outbound|8000||httpbin.bar.global
```
```
$ istioctl -n istio-system proxy-config endpoints istio-egressgateway-79b6467b8-g9d5t |grep httpbin
10.100.1.253:15443        HEALTHY     OK	outbound_.8000_._.httpbin.bar.global
10.100.1.253:15443        HEALTHY     OK	outbound|8000||httpbin.bar.global
101.102.114.161:15443     HEALTHY     OK	outbound_.8000_._.httpbin.bar.global
101.102.114.161:15443     HEALTHY     OK	outbound|8000||httpbin.bar.global
```
I am concerned about the external endpoint (`101.102.114.161`) set in the *sleep* proxy. As far as I understand, with the applied ServiceEntry policy, the proxy should see only the internal endpoint of the Egress Gateway (`10.100.1.253`). Is that correct?

I expected, when adding ServiceEntry with specified network:
```yaml
  Endpoints:
    Address:  101.102.114.161
    Network:  external
    Ports:
      http1:  15443
    Address:  10.100.1.253
    Ports:
      http1:  15443
```
... and specifying network view for Egress Gateway:
```
$ helm get values istio |grep -i network_view -C 5
gateways:
  istio-egressgateway:
    enabled: true
    env:
      ISTIO_META_REQUESTED_NETWORK_VIEW: external
    podAnnotations:
      openstack.org/kuryr-multi-vif-specs: '[{"type": "regular", "fixed_ips": [{"subnet_id":
        "a61b8ac6-1e3e-4376-9d71-011cd2f5c837"}]}]'
```
Then, only Egress Gateway would see the endpoint in the external network (` 101.102.114.161`).

**Comment by bzurkowski** [comment]
Potentially related issues: https://github.com/istio/istio/issues/12522, https://github.com/istio/istio/issues/3312, https://github.com/istio/istio/issues/12224.

**Comment by bzurkowski** [comment]
Interesting fact. After editing the ServiceEntry to use DNS strategy for address resolution and then providing address of Egress Gateway via FQDN, the proxy endpoints in *sleep* pod and Egress Gateway have been set correctly:
```
$ istioctl -n foo proxy-config endpoints sleep-f8cbf5b76-lhh9c |grep httpbin
10.100.1.253:15443     HEALTHY     OK                outbound|8000||httpbin.bar.global
```
```
$ istioctl -n istio-system proxy-config endpoints istio-egressgateway-79b6467b8-g9d5t |grep httpbin
101.102.114.161:15443     HEALTHY     OK                outbound_.8000_._.httpbin.bar.global
101.102.114.161:15443     HEALTHY     OK                outbound|8000||httpbin.bar.global
```
New ServiceEntry policy:
```yaml
Name:         httpbin-bar
Namespace:    foo
Labels:       <none>
API Version:  networking.istio.io/v1alpha3
Kind:         ServiceEntry
Metadata:
  Creation Timestamp:  2020-03-13T12:11:34Z
  Generation:          1
  Resource Version:    47682708
  Self Link:           /apis/networking.istio.io/v1alpha3/namespaces/foo/serviceentries/httpbin-bar
  UID:                 10bb9f03-01ec-4f50-abd8-d8538222f04a
Spec:
  Addresses:
    240.0.0.2
  Endpoints:
    Address:  101.102.114.161
    Network:  external
    Ports:
      http1:  15443
    Address:  istio-egressgateway.istio-system.svc.cluster.local
    Ports:
      http1:  15443
  Hosts:
    httpbin.bar.global
  Location:  MESH_INTERNAL
  Ports:
    Name:      http1
    Number:    8000
    Protocol:  http
  Resolution:  DNS
Events:        <none>
```
That solves the first issue, but the second one remains. I still get `503` response from the proxy:
```
HTTP/1.1 503 Service Unavailable
content-length: 91
content-type: text/plain
date: Thu, 19 Mar 2020 11:45:24 GMT
server: envoy
```

**Comment by bzurkowski** [comment]
I identified that this is the proxy inside *httpbin* pod in Cluster 2, responding with `400 Bad Request` to `TLS Client Hello` request from Egress Gateway in Cluster 1.

Any idea why the remote proxy would respond with `400` to TLS handshake?

**Comment by bzurkowski** [comment]
It might be related to HTTP-HTTPS mismatch: https://github.com/istio/istio/issues/12066#issuecomment-468480131, https://github.com/istio/istio/issues/8029

**Comment by bzurkowski** [comment]
Solved the issue. I was **missing default auth policy** in Cluster 2:
```
$ istioctl -n istio-system authn tls-check istio-ingressgateway-55898ccb8-c6xr2 httpbin.bar.svc.cluster.local
HOST:PORT                              STATUS     SERVER     CLIENT     AUTHN POLICY     DESTINATION RULE
httpbin.bar.svc.cluster.local:8000     OK         HTTP       HTTP       -                -
```
After creating the policy:
```
cat <<EOF | kubectl delete -n bar -f -
apiVersion: authentication.istio.io/v1alpha1
kind: MeshPolicy
metadata:
  name: default
spec:
  peers:
  - mtls:
      mode: PERMISSIVE
EOF
```
... *httpbin* app has been set to consume mTLS traffic:
```
$ istioctl -n istio-system authn tls-check istio-ingressgateway-55898ccb8-c6xr2 httpbin.bar.svc.cluster.local
HOST:PORT                              STATUS     SERVER        CLIENT     AUTHN POLICY     DESTINATION RULE
httpbin.bar.svc.cluster.local:8000     OK         HTTP/mTLS     HTTP       default/         -
```
Now, request from *sleep* pod succeeds:
```
$ kubectl exec --context=$CTX_CLUSTER1 $SLEEP_POD -n foo -c sleep -- curl -I httpbin.bar.global:8000/headers
HTTP/1.1 200 OK
server: envoy
date: Thu, 19 Mar 2020 13:22:15 GMT
content-type: application/json
content-length: 519
access-control-allow-origin: *
access-control-allow-credentials: true
x-envoy-upstream-service-time: 27
```
I guess, without the default mesh policy, mTLS is disabled?

**Comment by bzurkowski** [comment]
Anyway, the status of the issue is as follows:
> Envoy proxy in sleep pod attempts to call external Ingress Gateway directly instead of routing request via the Egress Gateway in the first place

Fixed by setting address resolution strategy to DNS and using FQDN for Egress Gateway's address. It must be checked, why the prior (STATIC) strategy didn't work - it is described in the documentation.

> Egress Gateway (or other remote service on the request path) responds with HTTP 400 Bad Request on TLS Client Hello.

Fixed by adding missing default authentication policy.

**Comment by kpiyush17** [comment]
@bzurkowski 

```
cat <<EOF | kubectl delete -n bar -f -
apiVersion: authentication.istio.io/v1alpha1
kind: MeshPolicy
metadata:
  name: default
spec:
  peers:
  - mtls:
      mode: PERMISSIVE
EOF
```
why kubectl delete?

This command does not work in istio-1.6.3:
`istioctl -n istio-system authn tls-check istio-ingressgateway-55898ccb8-c6xr2 httpbin.bar.svc.cluster.local`

Do you happen to know the lastest one?
